### ISSUE TITLE
start up

### ISSUE THREAD

**Issue by velociwabbit** [marmelab/admin-on-rest#284]
I have been using this very cool  mini framework for a few days and I am of two minds about it. One the idea is great and two ... who is the audience for your instructions?
For example

'make run' ?  to run the examples?  this is both cryptic and non standard and very frustrating for anyone who has a windows front end   environment which is likely to be 70% of the world. 

I have nmake and other build libraries on my computer but make is a unix specific and non-standard for node based applications.   Do i need to run my git console?  (I know the answer but really come on!)

Where is the npm start function to show something cool right away?

Also the 'hacking instructions are also fairly complex which is just yet again frustrating for no reason.

How can I help to make this a better solution?

**Comment by fzaninotto** [comment]
Hi,

The audience of the instructions is people who know and use `make`. If you're on Windows and don't use it, sorry, you're not in our target audience. 

Tip: In that case, open the `makefile` and copy the instructions by hand.

**Comment by velociwabbit** [comment]
would you like me to create a set of instructions for those who are forced
to use windows?

**Comment by fzaninotto** [comment]
Yep, if you think you can make something simple.

**Comment by velociwabbit** [comment]
will do.

I did finally get a configuration that works and have developed a brute
force two file (about 1k lines total) with all your modules in it. (Just to
make sure that I understand the app thoroughly )

I am still debugging it as there are some subtleties when combining modules
and removing the tree hierarchy.

 The biggest challenge I am having is to wade through all the method calls
in react / redux / saga to identify my consolidation errors.  Perhaps there
is a standard way to do this I am presently using chrome tools and react
tools for chrome but debugging .jsx by stepping through code to figure out
where my syntax/semantics are  wrong is non trivial.  Any suggestions would
be appreciated.

BTW there are some really elegant components to admin-on-rest. One of my
favorite is  withprops.js. It took awhile to figure out what it was doing
(=> semantics can be subtle ) but once i did it made me smile.

/jpr



On Wed, Feb 8, 2017 at 8:18 AM, Francois Zaninotto <notifications@github.com

**Comment by fzaninotto** [comment]
I don't understand what you're doing. Combining modules is the work of webpack or other module bundlers. Check for instance [admin-on-rest-demo](https://github.com/marmelab/admin-on-rest-demo) for an example usage.

**Comment by velociwabbit** [comment]
​Perhaps it is my limited abilities but I have found that ​there is
diminishing returns to modular software beyond a certain level of
node_module dependencies.  This is particularly true in a mission critical
application that must be supported by mere mortals.

Before adapting any library set I always  "de-modularize" most or all of
the code I utilize if it is new and unfamiliar. This presents me with a
simple and clean path through the code that helps to understand exactly how
it works.

It also give me an opportunity to learn how you code and what new and
interesting modules and ideas you use.

I have also found that for others who are learning the same library such a
stripped down version of the library helps them quickly understand what the
code does and how to immediately take advantage of it.

IMHO software development has become nearly baroque in its unnecessary
ornamentation and complexity . No doubt there are passable
computer-science-based arguments for making code larger and more complex
than the minimal alternative but once again mere mortals do not have the
time or disposition to embrace formal systems when informal systems are
easier to deal with and more fun to code.

What is happening to JavaScript with es6 is a perfect example of cs gone
wild.    Instead of learning how elegant prototype-based languages truly
are these same bone heads are trying to turn js into Java or some other
anal retentive/ nanny state language.

I have rewritten your entire library in about 900 LOC. in two modules (one
for instantiating my code and the other for the rest of your library.

I plan on reducing it down to  about 500 loc (or so) when i am finished.

If you like I will send you a copy  when I am done.

/jpr

**Comment by fzaninotto** [comment]
I share none of your concerns, but I'm curious to see how you've transformed a modular, tested, maintainable, extensible library into a 500 loc file *without* losing any of the qualities I just noted.

**Comment by velociwabbit** [comment]
You must work with some awesome developers if you are not concerned about
module proliferation and version management.

with regard to testing... your code is well tested already. All i do is
eliminate pathways and options that I know we will not use .

with  regard to maintaining the code.  I would assert that 500 loc is a lot
easier to debug and maintain then sorting out the same in 500 separate
modules ( including things like inference etc) all with at least 100 loc.

The motto for real estate and virtual real estate is the same  location
location location.

Google modified its default to provide 100 (instead of ~ 20) responses to a
query and found a drop off of ~50% utilization They attributed this to the
fact that response times dropped by nearly 3/4 of a second.  IBM did the
same study in early 80's and found the same phenomena.   People get
distracted and do other stuff.

Searching for modules and figuring out which ones are being used and when
is a time sump that offers the same type of frustration as waiting for a
response from a search engine.

On the other hand if one knows that the answer is in 500 loc they are more
likely to stay focused.

The problem with CS is that it assumes that everyone thinks and acts like
an academic who does not have ADD and usually possesses a very good short
term memory.

I have found that such people dont really innovate. They derive and as such
their systems  are turgid and boring which makes them hard to follow and
therefore understand.


When i read I prefer Hemingway to James Joyce for the same reason.


Joyce might be more precise but Hemingway  was much more compelling.

/jpr

**Comment by fzaninotto** [comment]
Dude, you sure like to speak. But we're not here to compare Joyce and Hemingway (I prefer Proust and Flaubert). Just show us some code.

**Comment by velociwabbit** [comment]
In due time as I said I have a bug and chrome dev tools is only showing it
failing in  react modules and the call stack does not include my calling
function in jsx which makes it  very hard to diagnose.

I am not a react expert which also makes it hard.

I have installed the react plug in but that does not place breakpoints in
the code.   I have the app working in non-consolidated mode but not as a
standalone app with only react/redux/sideeffects.


any suggestions?

/jpr

**Comment by fzaninotto** [comment]
Admin-on-rest currently contains about 6000 loc. If you've managed to reduce that to 500, you may consider patenting your method as a very efficient compression algorithm. Except, of course, I can also compare it to the compiled `index.js` when I use admin-on-rest with Webpack, which counts... 44 loc.

Let's end the discussion here: We can all agree that webpack is better than you at reducing locs.
